### PR TITLE
loading.tsx should have no effect on partial rendering when PPR is enabled

### PR DESCRIPTION
**Pull request by acdlite**
Before PPR, the way instant navigations work in Next.js is we prefetch everything up to the first route segment that defines a loading.js boundary. The rest of the tree is defered until the actual navigation. It does not take into account whether the data is dynamic — even if the tree is completely static, it will still defer everything inside the loading boundary.

The approach with PPR is different — we prefetch as deeply as possible, and only defer when dynamic data is accessed. If so, we only defer the nearest parent Suspense boundary of the dynamic data access, regardless of whether the boundary is defined by loading.js or a normal <Suspense> component in userspace.

This PR removes the partial behavior of loading.js when the PPR flag is enabled. In effect, loading.js now acts like a regular Suspense boundary with no additional special behavior.

Note that in practice this usually means we'll end up prefetching more than we were before PPR, which may or may not be considered a performance regression by some apps. The plan is to address this before General Availability of PPR by introducing granular per-segment fetching, so we can reuse as much of the tree as possible during both prefetches and dynamic navigations. But during the beta period, we should be clear about this trade off in our communications.

## Testing strategy

While I was writing a test, I noticed that it's currently pretty difficult to test all the scenarios that PPR is designed to handle, so I gave special attention to setting up a testing strategy that I hope will make this easier going forward. The overall pattern is based on how we've been testing concurrent rendering features in the React repo for many years:

- In the e2e test, spin up an HTTP server for responding to requests sent by the test app. This simulates the data service that would be used in a real Next.js application, whether it's direct db access, an ORM, or a higher-level data access layer. The e2e test can observe when individual requests are received, and control the timing of when the data is fulfilled, without needing to mock any lower level I/O. (We're already using a similar pattern to [test fetch deduping](https://github.com/vercel/next.js/blob/a3616d33edc40e7717b258f2636f280e30a3bccb/test/e2e/app-dir/app-fetch-deduping/app-fetch-deduping.test.ts#L8-L29).)
- Each time a request is received, write to an event log. Then assert on the result of the log at different points throughout the test. This helps catch subtle mistakes where the order of events is not expected, or the same event happens more than it should.

(I wrote some test helpers, but to avoid early abstraction, I've intentionally not moved them into a separate module.)

Closes NEXT-1779

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 8e5221da960802eba4cf55ccffb043fefb773a86

`pnpm test test/integration/clean-distdir/test/index.test.js` 
- Cleaning distDir > disabled write > should not clean up .next before build start

<details>
<summary>Expand output</summary>

  ● Cleaning distDir › disabled write › should not clean up .next before build start

    thrown: "Exceeded timeout of 60000 ms for a test.
    Add a timeout value to this test to increase the timeout, if this is a long-running test. See https://jestjs.io/docs/api#testname-fn-timeout."

      50 |     })
      51 |
    > 52 |     it('should not clean up .next before build start', async () => {
         |     ^
      53 |       await checkFileWrite(true)
      54 |     })
      55 |   })

      at it (integration/clean-distdir/test/index.test.js:52:5)
      at describe (integration/clean-distdir/test/index.test.js:36:3)
      at Object.describe (integration/clean-distdir/test/index.test.js:31:1)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | acdlite/next.js ppr-loading-tsx | Change |
| - | - | - | - |
| buildDuration | 10.7s | 10.8s | ⚠️ +112ms |
| buildDurationCached | 6s | 6.7s | ⚠️ +717ms |
| nodeModulesSize | 199 MB | 199 MB | ⚠️ +32.9 kB |
| nextStartRea..uration (ms) | 420ms | 423ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | acdlite/next.js ppr-loading-tsx | Change |
| - | - | - | - |
| 199-HASH.js gzip | 30.8 kB | 30.8 kB | N/A |
| 3f784ff6-HASH.js gzip | 53.3 kB | 53.3 kB | ✓ |
| 494.HASH.js gzip | 180 B | 181 B | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 241 B | 239 B | N/A |
| main-HASH.js gzip | 31.7 kB | 31.7 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | ✓ |
| Overall change | 100 kB | 100 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | acdlite/next.js ppr-loading-tsx | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | acdlite/next.js ppr-loading-tsx | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 195 B | N/A |
| _error-HASH.js gzip | 182 B | 181 B | N/A |
| amp-HASH.js gzip | 501 B | 503 B | N/A |
| css-HASH.js gzip | 322 B | 323 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | ✓ |
| edge-ssr-HASH.js gzip | 253 B | 255 B | N/A |
| head-HASH.js gzip | 348 B | 347 B | N/A |
| hooks-HASH.js gzip | 369 B | 368 B | N/A |
| image-HASH.js gzip | 4.27 kB | 4.27 kB | N/A |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.61 kB | 2.6 kB | N/A |
| routerDirect..HASH.js gzip | 311 B | 311 B | ✓ |
| script-HASH.js gzip | 384 B | 383 B | N/A |
| withRouter-HASH.js gzip | 307 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.17 kB | 3.17 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | acdlite/next.js ppr-loading-tsx | Change |
| - | - | - | - |
| _buildManifest.js gzip | 484 B | 483 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | acdlite/next.js ppr-loading-tsx | Change |
| - | - | - | - |
| index.html gzip | 529 B | 527 B | N/A |
| link.html gzip | 539 B | 542 B | N/A |
| withRouter.html gzip | 525 B | 522 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | acdlite/next.js ppr-loading-tsx | Change |
| - | - | - | - |
| edge-ssr.js gzip | 92.6 kB | 92.7 kB | N/A |
| page.js gzip | 145 kB | 145 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | acdlite/next.js ppr-loading-tsx | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 626 B | 624 B | N/A |
| middleware-r..fest.js gzip | 150 B | 151 B | N/A |
| middleware.js gzip | 35.7 kB | 35.7 kB | ✓ |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 37.6 kB | 37.6 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | acdlite/next.js ppr-loading-tsx | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 168 kB | 168 kB | N/A |
| app-page-exp..prod.js gzip | 93.8 kB | 93.8 kB | N/A |
| app-page-tur..prod.js gzip | 94.5 kB | 94.5 kB | N/A |
| app-page-tur..prod.js gzip | 89.1 kB | 89.1 kB | N/A |
| app-page.run...dev.js gzip | 138 kB | 138 kB | N/A |
| app-page.run..prod.js gzip | 88.4 kB | 88.4 kB | N/A |
| app-route-ex...dev.js gzip | 24.2 kB | 24.2 kB | ✓ |
| app-route-ex..prod.js gzip | 16.8 kB | 16.8 kB | ✓ |
| app-route-tu..prod.js gzip | 16.9 kB | 16.9 kB | ✓ |
| app-route-tu..prod.js gzip | 16.4 kB | 16.4 kB | ✓ |
| app-route.ru...dev.js gzip | 23.6 kB | 23.6 kB | ✓ |
| app-route.ru..prod.js gzip | 16.4 kB | 16.4 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.64 kB | 9.64 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-turbo...prod.js gzip | 21.9 kB | 21.9 kB | ✓ |
| pages.runtim...dev.js gzip | 22.6 kB | 22.6 kB | ✓ |
| pages.runtim..prod.js gzip | 21.9 kB | 21.9 kB | ✓ |
| server.runti..prod.js gzip | 49.3 kB | 49.3 kB | ✓ |
| Overall change | 258 kB | 258 kB | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-exp..ntime.dev.js</strong></summary>

```diff
failed to diff
```
</details>
<details>
<summary>Diff for <strong>app-page-exp..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page-tur..time.prod.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.dev.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>app-page.runtime.prod.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 8e5221da960802eba4cf55ccffb043fefb773a86

File: test/e2e/app-dir/ppr-navigations/ppr-navigations.test.ts [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+    // during initial render, it still acts as a regular Suspense boundary.
+    // Trigger a "bad" Suspense fallback by intentionally suspending without
+    // startTransition.
+    await browser.elementById('trigger-bad-suspense-fallback').click()
+    const loading = await browser.elementById('loading-tsx')
+    expect(await loading.innerHTML()).toEqual('Loading [inner loading.tsx]...')
+  })
+})
+
+// NOTE: I've intentionally not yet moved these helpers into a separate
```
Comment (ztanner) [inline test/e2e/app-dir/ppr-navigations/ppr-navigations.test.ts:99]: Should these things be part of `test-data-service` or is it intentionally separate?

Comment (acdlite) [reply test/e2e/app-dir/ppr-navigations/ppr-navigations.test.ts:99]: Eventually I'll probably move them to something like `test-data-service/writer` and `test-data-service/reader` but I just kept the `writer` ones in here for now

**Review by ztanner — APPROVED** [review]

**Pull request merged**